### PR TITLE
[#2979] feat(spark-connector): Implement FunctionCatalog in GravitinoIcebergCatalog

### PR DESCRIPTION
**Pull request by caican00**
### What changes were proposed in this pull request?
Implement `FunctionCatalog` in `GravitinoIcebergCatalog`

### Why are the changes needed?

Implement `FunctionCatalog` in `GravitinoIcebergCatalog` to support `Iceberg` partition functions.

Fix: https://github.com/datastrato/gravitino/issues/2979

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
New ITs.

### CODE REVIEW

File: integration-test/src/test/java/com/datastrato/gravitino/integration/test/spark/iceberg/SparkIcebergCatalogIT.java [hunk 1]
```diff
@@ -55,4 +63,68 @@ void testIcebergFileLevelDeleteOperation() {
...
+    functions.forEach(
+        function -> {
+          Identifier identifier = Identifier.of(new String[] {getDefaultDatabase()}, function);
+          Assertions.assertThrows(
+              NoSuchFunctionException.class, () -> functionCatalog.loadFunction(identifier));
+        });
+  }
+
+  @Test
+  void testCallingFunctionInSQLEndToEnd() {
```
Comment (FANNG1) [inline integration-test/src/test/java/com/datastrato/gravitino/integration/test/spark/iceberg/SparkIcebergCatalogIT.java:109]: rename to testIcebergFunction?

File: integration-test/src/test/java/com/datastrato/gravitino/integration/test/spark/iceberg/SparkIcebergCatalogIT.java [hunk 2]
```diff
@@ -55,4 +63,68 @@ void testIcebergFileLevelDeleteOperation() {
     List<String> queryResult2 = getTableData(tableName);
     Assertions.assertEquals(0, queryResult2.size());
   }
+
+  @Test
+  void testFunctionsViaGravitinoIcebergCatalog()
```
Comment (FANNG1) [inline integration-test/src/test/java/com/datastrato/gravitino/integration/test/spark/iceberg/SparkIcebergCatalogIT.java:68]: rename to testIcebergListFunctions?

**Comment by FANNG1** [comment]
LGTM, except minor comments

File: spark-connector/spark-connector/src/main/java/com/datastrato/gravitino/spark/connector/iceberg/GravitinoIcebergCatalog.java [hunk 3]
```diff
@@ -74,6 +78,26 @@ protected PropertiesConverter getPropertiesConverter() {
     return new IcebergPropertiesConverter();
   }
 
+  @Override
+  public Identifier[] listFunctions(String[] namespace) throws NoSuchNamespaceException {
+    Preconditions.checkArgument(
```
[the remark this answers is not in the document]
Comment (caican00) [reply spark-connector/spark-connector/src/main/java/com/datastrato/gravitino/spark/connector/iceberg/GravitinoIcebergCatalog.java]: Already removed. thanks

File: integration-test/src/test/java/com/datastrato/gravitino/integration/test/spark/iceberg/SparkIcebergCatalogIT.java (hunk 2 above)
Comment (caican00) [reply integration-test/src/test/java/com/datastrato/gravitino/integration/test/spark/iceberg/SparkIcebergCatalogIT.java]: fixed.

File: integration-test/src/test/java/com/datastrato/gravitino/integration/test/spark/iceberg/SparkIcebergCatalogIT.java (hunk 1 above)
Comment (caican00) [reply integration-test/src/test/java/com/datastrato/gravitino/integration/test/spark/iceberg/SparkIcebergCatalogIT.java]: fixed.

**Comment by caican00** [comment]
> LGTM, except minor comments

all comments have been addressed. cc @FANNG1

**Review by FANNG1 — APPROVED** [review]

**Comment by FANNG1** [comment]
@caican00 , merged to main, thanks for your work!

**Comment by caican00** [comment]
> @caican00 , merged to main, thanks for your work!

@FANNG1 thanks for your review and merge!

**Pull request merged**